storage/paths/db_index.go

```diff
@@ -0,0 +1,891 @@
+		_, err := dbi.harmonyDB.BeginTransaction(ctx, func(tx *harmonydb.Tx) (commit bool, err error) {
+			// Drop storage path completely
+			_, err = dbi.harmonyDB.Exec(ctx, "DELETE FROM storagelocation WHERE storage_id=$1", id)
+			if err != nil {
+				return false, err
+			}
+
+			// Drop all sectors entries which use this storage path
+			_, err = dbi.harmonyDB.Exec(ctx, "DELETE FROM sectorlocation WHERE storage_id=$1", id)
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		})
+		if err != nil {
+			return err
+		}
+		log.Warnw("Dropping sector storage", "path", id)
+	}
+
+	return nil
+}
+
+func (dbi *DBIndex) StorageReportHealth(ctx context.Context, id storiface.ID, report storiface.HealthReport) error {
+
+	var canSeal, canStore bool
+	err := dbi.harmonyDB.QueryRow(ctx,
+		"SELECT can_seal, can_store FROM storagelocation WHERE storage_id=$1", id).Scan(&canSeal, &canStore)
+	if err != nil {
+		return xerrors.Errorf("Querying for storage id %s fails with err %v", id, err)
+	}
+
+	_, err = dbi.harmonyDB.Exec(ctx,
+		"UPDATE storagelocation set capacity=$1, available=$2, fs_available=$3, reserved=$4, used=$5, last_heartbeat=$6",
+		report.Stat.Capacity,
+		report.Stat.Available,
+		report.Stat.FSAvailable,
+		report.Stat.Reserved,
+		report.Stat.Used,
+		time.Now())
+	if err != nil {
+		return xerrors.Errorf("updating storage health in DB fails with err: %v", err)
+	}
+
+	if report.Stat.Capacity > 0 {
+		ctx, _ = tag.New(ctx,
+			tag.Upsert(metrics.StorageID, string(id)),
+			tag.Upsert(metrics.PathStorage, fmt.Sprint(canStore)),
+			tag.Upsert(metrics.PathSeal, fmt.Sprint(canSeal)),
+		)
+
+		stats.Record(ctx, metrics.StorageFSAvailable.M(float64(report.Stat.FSAvailable)/float64(report.Stat.Capacity)))
+		stats.Record(ctx, metrics.StorageAvailable.M(float64(report.Stat.Available)/float64(report.Stat.Capacity)))
+		stats.Record(ctx, metrics.StorageReserved.M(float64(report.Stat.Reserved)/float64(report.Stat.Capacity)))
+
+		stats.Record(ctx, metrics.StorageCapacityBytes.M(report.Stat.Capacity))
+		stats.Record(ctx, metrics.StorageFSAvailableBytes.M(report.Stat.FSAvailable))
+		stats.Record(ctx, metrics.StorageAvailableBytes.M(report.Stat.Available))
+		stats.Record(ctx, metrics.StorageReservedBytes.M(report.Stat.Reserved))
+
+		if report.Stat.Max > 0 {
+			stats.Record(ctx, metrics.StorageLimitUsed.M(float64(report.Stat.Used)/float64(report.Stat.Max)))
+			stats.Record(ctx, metrics.StorageLimitUsedBytes.M(report.Stat.Used))
+			stats.Record(ctx, metrics.StorageLimitMaxBytes.M(report.Stat.Max))
+		}
+	}
+
+	return nil
+}
+
+func (dbi *DBIndex) StorageDeclareSector(ctx context.Context, storageID storiface.ID, s abi.SectorID, ft storiface.SectorFileType, primary bool) error {
+
+	ftValid := false
+	for _, fileType := range storiface.PathTypes {
+		if fileType&ft == 0 {
+			ftValid = true
+			break
+		}
+	}
+	if !ftValid {
+		return xerrors.Errorf("Invalid filetype")
+	}
```

Comment:
* Would be nice to dedupe this with the same bit of code below
* Could call `ft.AllSet() -> []SectorFileType` and iterate over that
  * Or do asserts on that, at least reduces the direct bit fiddling here